### PR TITLE
librgw: add ability to conditionally export HTTP frontends/apis

### PR DESCRIPTION
**Pull request by mattbenjamin**
Signed-off-by: Matt Benjamin <mbenjamin@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by dang — APPROVED** [review]

**Comment by mattbenjamin** [comment]
```
[       OK ] LibRGW.SHUTDOWN (5950 ms)
[----------] 24 tests from LibRGW (12258 ms total)

[----------] Global test environment tear-down
[==========] 24 tests from 1 test suite ran. (12258 ms total)
[  PASSED  ] 24 tests.
```

**Comment by mattbenjamin** [comment]
hi or, when you get a chance, could you look at how I might be abusing ActiveRateLimiter here?  As far as I can tell, everything is working, but I'm getting an exception when RGWLib::ratelimiter is deleted:
`[----------] Global test environment tear-down
[==========] 24 tests from 1 test suite ran. (1490 ms total)
[  PASSED  ] 24 tests.
terminate called after throwing an instance of 'std::system_error'
  what():  Invalid argument
*** Caught signal (Aborted) **
 in thread 7fe256181dc0 thread_name:ceph_test_librg
 ceph version 17.0.0-14446-g96f16f514a2 (96f16f514a2ddbb2332e1f61006e96ae19b39fb0) quincy (dev)
 1: /home/mbenjamin/ceph-cp/build/lib/librgw.so.2(+0x17ed13b) [0x7fe25bd3e13b]
 2: /lib64/libpthread.so.0(+0x14a90) [0x7fe2593dea90]
 3: gsignal()
 4: abort()
 5: /lib64/libstdc++.so.6(+0xaac19) [0x7fe258ccac19]
 6: /lib64/libstdc++.so.6(+0xb62fa) [0x7fe258cd62fa]
 7: /lib64/libstdc++.so.6(+0xb5369) [0x7fe258cd5369]
 8: __gxx_personality_v0()
 9: /lib64/libgcc_s.so.1(+0x10bff) [0x7fe258acebff]
 10: _Unwind_RaiseException()
 11: __cxa_throw()
 12: (std::__throw_system_error(int)+0x85) [0x7fe258ccd8ec]
 13: (std::thread::detach()+0) [0x7fe258d01fa0]
 14: (ActiveRateLimiter::~ActiveRateLimiter()+0x12e) [0x7fe25b672bf4]
 15: (ActiveRateLimiter::~ActiveRateLimiter()+0x9) [0x7fe25b672c85]
 16: (std::unique_ptr<ActiveRateLimiter, std::default_delete<ActiveRateLimiter> >::~unique_ptr()+0x12) [0x7fe25b6a116c]
 17: (rgw::RGWLib::~RGWLib()+0x5b) [0x7fe25b6a249b]
 18: __cxa_finalize()
 19: /home/mbenjamin/ceph-cp/build/lib/librgw.so.2(+0x10feca7) [0x7fe25b64fca7]
Aborted (core dumped)
`

**Comment by mattbenjamin** [comment]
also passing when the store backend is dbstore:
```[mbenjamin@lemon build]$ ./bin/ceph_test_librgw_file_nfsns  --hier1 --dirs1 --create --rename --verbose
[==========] Running 24 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 24 tests from LibRGW
[ RUN      ] LibRGW.INIT
[       OK ] LibRGW.INIT (103 ms)
[ RUN      ] LibRGW.MOUNT_NOROOT
[       OK ] LibRGW.MOUNT_NOROOT (0 ms)
[ RUN      ] LibRGW.MOUNT
[       OK ] LibRGW.MOUNT (0 ms)
[ RUN      ] LibRGW.SETUP_HIER1
creating: nfsroot:foo/bar/baz/quux
creating: nfsroot:foo/f1
creating: nfsroot:foo/f2
creating: nfsroot:foo/bar/f1
creating: nfsroot:foo/bar/d1/
creating: nfsroot:foo/bar/baz/hungry
creating: nfsroot:foo/bar/baz/hungry/
creating: nfsroot:foo/bar/baz/momma
creating: nfsroot:foo/bar/baz/bear/
creating: nfsroot:foo/bar/baz/sasquatch
creating: nfsroot:foo/bar/baz/sasquatch/
creating: nfsroot:foo/bar/baz/frobozz
[       OK ] LibRGW.SETUP_HIER1 (19 ms)
[ RUN      ] LibRGW.SETUP_DIRS1
[       OK ] LibRGW.SETUP_DIRS1 (48 ms)
[ RUN      ] LibRGW.SETATTR
[       OK ] LibRGW.SETATTR (0 ms)
[ RUN      ] LibRGW.RGW_CREATE_DIRS1
[       OK ] LibRGW.RGW_CREATE_DIRS1 (8 ms)
[ RUN      ] LibRGW.RGW_SETUP_RENAME1
[       OK ] LibRGW.RGW_SETUP_RENAME1 (21 ms)
[ RUN      ] LibRGW.RGW_INTRABUCKET_RENAME1
rename file rfile_0 to rfile_r0 (bucket brename0)
[       OK ] LibRGW.RGW_INTRABUCKET_RENAME1 (2 ms)
[ RUN      ] LibRGW.RGW_CROSSBUCKET_RENAME1
rename file rfile_1 (bucket brename0) to rfile_rhilldog (bucket brename1)
[       OK ] LibRGW.RGW_CROSSBUCKET_RENAME1 (2 ms)
[ RUN      ] LibRGW.GETATTR_DIRS1
scanning objects in /bdirs1/dir_0
	/bdirs1/dir_0/sdir_0 (sdir_0): DIR  mode: 17139 nlinks: 1
	/bdirs1/dir_0/sfile_0 (sfile_0): FILE  mode: 33412 size: 16
	/bdirs1/dir_0/sdir_1 (sdir_1): DIR  mode: 17139 nlinks: 1
	/bdirs1/dir_0/sfile_1 (sfile_1): FILE  mode: 33412 size: 16
scanning objects in /bdirs1/dir_1
	/bdirs1/dir_1/sdir_0 (sdir_0): DIR  mode: 17139 nlinks: 1
	/bdirs1/dir_1/sfile_0 (sfile_0): FILE  mode: 33412 size: 16
	/bdirs1/dir_1/sdir_1 (sdir_1): DIR  mode: 17139 nlinks: 1
	/bdirs1/dir_1/sfile_1 (sfile_1): FILE  mode: 33412 size: 16
scanning objects in /bdirs1/dir_2
	/bdirs1/dir_2/sdir_0 (sdir_0): DIR  mode: 17139 nlinks: 1
	/bdirs1/dir_2/sfile_0 (sfile_0): FILE  mode: 33412 size: 16
	/bdirs1/dir_2/sdir_1 (sdir_1): DIR  mode: 17139 nlinks: 1
	/bdirs1/dir_2/sfile_1 (sfile_1): FILE  mode: 33412 size: 16
[       OK ] LibRGW.GETATTR_DIRS1 (0 ms)
[ RUN      ] LibRGW.READ_DIRS1
read back objects in /bdirs1/dir_0
reading 0,256 dir_0/sfile_0
	read back from sfile_0 : "data for sfile_0"
reading 0,256 dir_0/sfile_1
	read back from sfile_1 : "data for sfile_1"
read back objects in /bdirs1/dir_1
reading 0,256 dir_1/sfile_0
	read back from sfile_0 : "data for sfile_0"
reading 0,256 dir_1/sfile_1
	read back from sfile_1 : "data for sfile_1"
read back objects in /bdirs1/dir_2
reading 0,256 dir_2/sfile_0
	read back from sfile_0 : "data for sfile_0"
reading 0,256 dir_2/sfile_1
	read back from sfile_1 : "data for sfile_1"
[       OK ] LibRGW.READ_DIRS1 (4 ms)
[ RUN      ] LibRGW.READF_SETUP1
[       OK ] LibRGW.READF_SETUP1 (272 ms)
[ RUN      ] LibRGW.READF_DIRS1
[       OK ] LibRGW.READF_DIRS1 (0 ms)
[ RUN      ] LibRGW.WRITEF_DIRS1
[       OK ] LibRGW.WRITEF_DIRS1 (0 ms)
[ RUN      ] LibRGW.RELEASE_DIRS1
release sdir_0 type: DIR refs: 2
release sfile_0 type: FILE refs: 2
release sdir_1 type: DIR refs: 2
release sfile_1 type: FILE refs: 2
release sdir_0 type: DIR refs: 2
release sfile_0 type: FILE refs: 2
release sdir_1 type: DIR refs: 2
release sfile_1 type: FILE refs: 2
release sdir_0 type: DIR refs: 2
release sfile_0 type: FILE refs: 2
release sdir_1 type: DIR refs: 2
release sfile_1 type: FILE refs: 2
[       OK ] LibRGW.RELEASE_DIRS1 (0 ms)
[ RUN      ] LibRGW.HIER1
/nfsroot/foo/f2 (f2): FILE 
/nfsroot/foo/f1 (f1): FILE 
/nfsroot/foo/bar/f1 (f1): FILE 
/nfsroot/foo/bar/d1 (d1): DIR 
/nfsroot/foo/bar/baz/sasquatch (sasquatch): FILE 
/nfsroot/foo/bar/baz/sasquatch (sasquatch): FILE 
/nfsroot/foo/bar/baz/quux (quux): FILE 
/nfsroot/foo/bar/baz/momma (momma): FILE 
/nfsroot/foo/bar/baz/hungry (hungry): FILE 
/nfsroot/foo/bar/baz/hungry (hungry): FILE 
/nfsroot/foo/bar/baz/frobozz (frobozz): FILE 
/nfsroot/foo/bar/baz/bear (bear): DIR 
/nfsroot/foo/bar/baz (baz): DIR 
/nfsroot/foo/bar (bar): DIR 
/nfsroot/foo (foo): DIR 
/nfsroot (nfsroot): DIR 
[       OK ] LibRGW.HIER1 (16 ms)
[ RUN      ] LibRGW.MARKER1_SETUP_BUCKET
[       OK ] LibRGW.MARKER1_SETUP_BUCKET (0 ms)
[ RUN      ] LibRGW.MARKER1_SETUP_OBJECTS
[       OK ] LibRGW.MARKER1_SETUP_OBJECTS (0 ms)
[ RUN      ] LibRGW.MARKER1_READDIR
[       OK ] LibRGW.MARKER1_READDIR (0 ms)
[ RUN      ] LibRGW.MARKER1_OBJ_CLEANUP
[       OK ] LibRGW.MARKER1_OBJ_CLEANUP (0 ms)
[ RUN      ] LibRGW.CLEANUP
[       OK ] LibRGW.CLEANUP (0 ms)
[ RUN      ] LibRGW.UMOUNT
[       OK ] LibRGW.UMOUNT (0 ms)
[ RUN      ] LibRGW.SHUTDOWN
2022-08-26T09:59:04.002-0400 7f81177fedc0 -1 shutting down
2022-08-26T09:59:04.002-0400 7f81097fa700 -1 received  signal: User defined signal 1, si_code : -6, si_value (int): 0, si_value (ptr): 0, si_errno: 0, si_pid : 686271, si_uid : 1001, si_addr0x3e9000a78bf, si_status0
[       OK ] LibRGW.SHUTDOWN (109 ms)
[----------] 24 tests from LibRGW (605 ms total)

[----------] Global test environment tear-down
[==========] 24 tests from 1 test suite ran. (605 ms total)
[  PASSED  ] 24 tests.
[mbenjamin@lemon build]$ /home/mbenjamin/ceph-cp/build/bin/radosgw -c /home/mbenjamin/ceph-cp/build/ceph.conf --log-file=/home/mbenjamin/ceph-cp/build/out/radosgw.8000.log --admin-socket=/home/mbenjamin/ceph-cp/build/out/radosgw.8000.asok --pid-file=/home/mbenjamin/ceph-cp/build/out/radosgw.8000.pid --rgw_luarocks_location=/home/mbenjamin/ceph-cp/build/out/luarocks --debug-rgw=20 --debug-ms=0 -n client.rgw.8000 --rgw_frontends='beast port=8000'
[mbenjamin@lemon build]$ s3cmd ls
2022-08-26 13:59  s3://bdirs1
2022-08-26 13:59  s3://brename0
2022-08-26 13:59  s3://brename1
2022-08-26 13:59  s3://nfsroot
2022-08-26 13:55  s3://wendy1
```

**Comment by mattbenjamin** [comment]
I still need to adjust the processing of rgw_frontends by librgw, but it's now testable.

**Comment by mattbenjamin** [comment]
@dang unified librgw and main(...), hopefully without omissions

**Comment by dang** [comment]
LGTM

File: doc/radosgw/nfs.rst [hunk 1]
```diff
@@ -122,13 +122,17 @@ Required ceph.conf configuration for RGW NFS includes:
...
-optional and in some cases ignored.
+Other config variables (e.g., ``rgw data`` and ``rgw backend store``) are
+optional.
 
 A small number of config variables (e.g., ``rgw_nfs_namespace_expire_secs``)
 are unique to RGW NFS.
 
+In particular, front-end selection is handled specially by the librgw.so runtime.  By default, only the
+``rgw-nfs`` frontend is started. Additional frontends (e.g., ``beast``) are enabled via the
+``rgw nfs frontends`` config option.  It's syntax is identical to the ordinary ``rgw frontends`` option.
```
Comment (anthonyeleven) [inline doc/radosgw/nfs.rst:133]: `its`, but better yet, but better yet take out `It’s` and capitalize Syntax.  “Syntax is identical to that of the ordinary…”

File: doc/radosgw/nfs.rst [hunk 2]
```diff
@@ -122,13 +122,17 @@ Required ceph.conf configuration for RGW NFS includes:
...
+Default options for non-default frontends are specified via ``rgw frontend defaults`` as normal.
+
```
Comment (anthonyeleven) [inline doc/radosgw/nfs.rst:135]: s/normal/usual/

File: src/common/options/rgw.yaml.in [hunk 3]
```diff
@@ -1148,6 +1148,22 @@ options:
   services:
   - rgw
   with_legacy: true
+- name: rgw_nfs_frontends
+  type: str
+  level: basic
+  desc: RGW frontends configuration when running as librgw/nfs
+  long_desc: A comma delimited list of frontends configuration. Each configuration
```
Comment (anthonyeleven) [inline src/common/options/rgw.yaml.in:1155]: Comma-delimited , below space-delimited please

File: src/rgw/rgw_main.h [hunk 4]
```diff
@@ -0,0 +1,130 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2022 New Dream Network
```
Comment (anthonyeleven) [inline src/rgw/rgw_main.h:7]: New dream network?

**Review by anthonyeleven — CHANGES_REQUESTED** [review]

File: doc/radosgw/nfs.rst (hunk 1 above)
Comment (mattbenjamin) [reply doc/radosgw/nfs.rst:133]: ok

File: src/common/options/rgw.yaml.in (hunk 3 above)
Comment (mattbenjamin) [reply src/common/options/rgw.yaml.in:1155]: ok

File: src/rgw/rgw_main.h (hunk 4 above)
Comment (mattbenjamin) [reply src/rgw/rgw_main.h:7]: ooh, good catch

**Comment by mattbenjamin** [comment]
@anthonyeleven updated, thanks for the review

**Review by anthonyeleven — APPROVED** [review]

File: src/rgw/CMakeLists.txt [hunk 5]
```diff
@@ -348,6 +348,8 @@ target_link_libraries(rgw_schedulers
 
 set(radosgw_srcs
   rgw_main.cc
+  rgw_file.cc
+  librgw.cc
```
Comment (cbodley) [inline src/rgw/CMakeLists.txt:352]: why are these compiled into the `radosgw` binary now?

File: src/common/options/rgw.yaml.in [hunk 6]
```diff
@@ -1148,6 +1148,22 @@ options:
   services:
   - rgw
   with_legacy: true
+- name: rgw_nfs_frontends
```
Comment (cbodley) [inline src/common/options/rgw.yaml.in:1151]: can you please add a `see_also` for `rgw_frontends`? the docs in https://docs.ceph.com/en/latest/radosgw/config-ref/ will add links

File: src/rgw/librgw.cc [hunk 7]
```diff
@@ -513,40 +508,13 @@ namespace rgw {
...
-					 run_gc,
-					 run_lc,
-					 run_quota,
-					 run_sync,
-					 g_conf().get_val<bool>("rgw_dynamic_resharding"));
+    common_init_finish(g_ceph_context);
 
-    if (!store) {
+    main.init_storage();
+    if (! main.store) {
```
Comment (cbodley) [inline src/rgw/librgw.cc:517]: this access to member variable `main.store` seems to be the only reason `RGWLib` needs to be a friend of `AppMain`, can you use `main.get_store()` instead?

File: src/rgw/rgw_main.h [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    std::unique_ptr<RGWRealmReloader> reloader;
+    std::unique_ptr<RGWPeriodPusher> pusher;
+    std::unique_ptr<RGWFrontendPauser> fe_pauser;
+    std::unique_ptr<RGWRealmWatcher> realm_watcher;
+    std::unique_ptr<RGWPauser> rgw_pauser;
+    rgw::sal::Store* store;
+    DoutPrefixProvider* dpp;
+
+    friend int main(int, char*[]); // doesnt work gcc-12.1
+    friend class RGWLib;
```
Comment (cbodley) [inline src/rgw/rgw_main.h:82]: can we get rid of the friendships here?

**Review by cbodley — COMMENTED** [review]
good stuff

File: src/rgw/CMakeLists.txt (hunk 5 above)
Comment (mattbenjamin) [reply src/rgw/CMakeLists.txt:352]: circular dependencies--I'll revisit and see if there's an alternative

File: src/common/options/rgw.yaml.in (hunk 6 above)
Comment (mattbenjamin) [reply src/common/options/rgw.yaml.in:1151]: sure

File: src/rgw/librgw.cc (hunk 7 above)
Comment (mattbenjamin) [reply src/rgw/librgw.cc:517]: sure

File: src/rgw/rgw_main.h (hunk 8 above)
Comment (mattbenjamin) [reply src/rgw/rgw_main.h:82]: sure

File: src/rgw/CMakeLists.txt (hunk 5 above)
Comment (cbodley) [reply src/rgw/CMakeLists.txt:352]: ah. would it help to move AppMain out of rgw_main.cc?

i think we should avoid linking a main() function into librgw

Comment (mattbenjamin) [reply src/rgw/CMakeLists.txt:352]: most likely;  it just seems inapropos, but most likely rgw_appmain.cc would do it

**Comment by mattbenjamin** [comment]
jenkins test make check

File: src/rgw/CMakeLists.txt [hunk 9]
```diff
@@ -270,33 +270,41 @@ if(WITH_RADOSGW_MOTR)
 endif()
 
 set(rgw_a_srcs
+  librgw.cc
```
Comment (cbodley) [inline src/rgw/CMakeLists.txt:273]: why do we need librgw.cc in both rgw_a_srcs and librgw_srcs? if it's necessary here in rgw_a, maybe we don't need a separate rgw target, and just make rgw_a the shared lib?

**Review by cbodley — COMMENTED** [review]
very nice cleanups in rgw_main 👍

File: src/rgw/rgw_signal.h [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ */
+
+#pragma once
+
+
+namespace rgw {
+  namespace signal {
+
+    void signal_shutdown();
```
[the remark this answers is not in the document]
Comment (mattbenjamin) [reply src/rgw/rgw_signal.h:22]: ok, easy to change--should I?

File: src/rgw/CMakeLists.txt (hunk 9 above)
Comment (mattbenjamin) [reply src/rgw/CMakeLists.txt:273]: it's only supposed to be in rgw_a now, I believe

Comment (mattbenjamin) [reply src/rgw/CMakeLists.txt:273]: I don't want to change what libs are built, in this PR, if that's ok--I have no specific objection, but I might have if I remembered why we have both

Comment (cbodley) [reply src/rgw/CMakeLists.txt:273]: okay, if we need it in rgw_a can you please remove librgw.cc from librgw_srcs so we're not compiling it twice?

File: src/rgw/rgw_signal.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (cbodley) [reply src/rgw/rgw_signal.h:22]: yes please, in rgw_main.h also

File: src/rgw/CMakeLists.txt (hunk 9 above)
Comment (mattbenjamin) [reply src/rgw/CMakeLists.txt:273]: right, agree, that's not intended

**Comment by mattbenjamin** [comment]
also, will fix this
```
/radosgw.dir/rgw_main.cc.o -MF src/rgw/CMakeFiles/radosgw.dir/rgw_main.cc.o.d -o src/rgw/CMakeFiles/radosgw.dir/rgw_main.cc.o -c /home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_main.cc
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_main.cc:27:37: error: unused variable 'rgw_op_tracepoint_traits' [-Werror,-Wunused-variable]
  extern TracepointProvider::Traits rgw_op_tracepoint_traits;
                                    ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_main.cc:28:37: error: unused variable 'rgw_rados_tracepoint_traits' [-Werror,-Wunused-variable]
  extern TracepointProvider::Traits rgw_rados_tracepoint_traits;
                                    ^
2 errors generated.
```

**Review by cbodley — APPROVED** [review]

**Comment by mattbenjamin** [comment]
https://pulpito.ceph.com/mbenjamin-2022-09-03_02:37:17-rgw-wip-librgw-frontends-distro-default-smithi/

**Comment by mattbenjamin** [comment]
https://pulpito.ceph.com/mbenjamin-2022-09-07_01:04:52-rgw-wip-librgw-frontends-distro-default-smithi/

**Comment by mattbenjamin** [comment]
https://pulpito.ceph.com/mbenjamin-2022-09-07_16:41:19-rgw-wip-librgw-frontends-distro-default-smithi/

**Comment by mattbenjamin** [comment]
https://pulpito.ceph.com/mbenjamin-2022-09-08_00:31:57-rgw-wip-librgw-frontends-distro-default-smithi/

**Comment by mattbenjamin** [comment]
https://pulpito.ceph.com/mbenjamin-2022-09-08_14:06:40-rgw-wip-librgw-frontends-distro-default-smithi/

File: src/rgw/rgw_lib.h [hunk 11]
```diff
@@ -25,35 +20,32 @@ namespace rgw {
   class RGWLibFrontend;
 
   class RGWLib : public DoutPrefixProvider {
-    RGWFrontendConfig* fec;
+    AppMain main;
```
Comment (cbodley) [inline src/rgw/rgw_lib.h:23]: `RGWLib` does hold a reference on `boost::intrusive_ptr<CephContext> cct` below, but on `RGWLib`'s destruction `cct` will be destroyed before `AppMain` is. should be able to reorder the member variables so `cct` comes first

Comment (mattbenjamin) [reply src/rgw/rgw_lib.h:23]: ok, thanks, that helps

Comment (mattbenjamin) [reply src/rgw/rgw_lib.h:23]: ok.  I'm cleaning up from a messy merge around dan's changes, will update this also

File: src/rgw/rgw_appmain.cc [hunk 12]
```diff
@@ -0,0 +1,608 @@
...
+  g_conf().remove_observer(implicit_tenant_context.get());
+  implicit_tenant_context.reset(); // deletes
+#ifdef WITH_RADOSGW_AMQP_ENDPOINT
+  rgw::amqp::shutdown();
+#endif
+#ifdef WITH_RADOSGW_KAFKA_ENDPOINT
+  rgw::kafka::shutdown();
+#endif
+  rgw_perf_stop(g_ceph_context);
+  ratelimiter.reset(); // deletes--ensure this happens before we destruct
```
[the remark this answers is not in the document]
Comment (mattbenjamin) [reply src/rgw/rgw_appmain.cc:607]: ok. I'm cleaning up from a messy merge around dan's changes, will update this also

**Comment by mattbenjamin** [comment]
https://pulpito.ceph.com/mbenjamin-2022-09-09_22:09:46-rgw-wip-librgw-frontends-distro-default-smithi/

**Comment by cbodley** [comment]
> Do you recognize these, correcting for some change in startup stack?
> https://www.klgrth.io/paste/nrhj6

i hadn't seen that before, no. but it doesn't look like your PR is causing it

a rados client is still running after `rgw::notify::Manager::~Manager()`, and it tries to schedule a librados completion on the `Manager`'s `io_context`. this would be the `librados::Rados` instance owned by `RGWSI_Rados`. cc @yuvalif

`RGWRados::finalize()` calls `shutdown()` on the services before `rgw::notify::shutdown()`: https://github.com/ceph/ceph/blob/c49b81c7d619cea23e9707d1f5bcc7de3049c4fd/src/rgw/rgw_rados.cc#L1091-L1109

but  `RGWSI_RADOS::shutdown()` doesn't actually shut down that rados client. i've opened https://github.com/ceph/ceph/pull/48041 to try that

**Comment by yuvalif** [comment]
> > Do you recognize these, correcting for some change in startup stack?
> > https://www.klgrth.io/paste/nrhj6
> 
> i hadn't seen that before, no. but it doesn't look like your PR is causing it
> 
> a rados client is still running after `rgw::notify::Manager::~Manager()`, and it tries to schedule a librados completion on the `Manager`'s `io_context`. this would be the `librados::Rados` instance owned by `RGWSI_Rados`. cc @yuvalif
> 
> `RGWRados::finalize()` calls `shutdown()` on the services before `rgw::notify::shutdown()`:
> 
> https://github.com/ceph/ceph/blob/c49b81c7d619cea23e9707d1f5bcc7de3049c4fd/src/rgw/rgw_rados.cc#L1091-L1109
> 
> but `RGWSI_RADOS::shutdown()` doesn't actually shut down that rados client. i've opened #48041 to try that

in the Manager's dtor I call `io_context.stop();` and also wait for all of the workers to finish their work.
how is more work being scheduled?

**Comment by mattbenjamin** [comment]
> 

I didn't think so, great work tracking this down so quickly.  Would you agree this PR is now ready?

**Comment by mattbenjamin** [comment]
jenkins test make check

**Comment by mattbenjamin** [comment]
jenkins test api

**Comment by cbodley** [comment]
> in the Manager's dtor I call `io_context.stop();` and also wait for all of the workers to finish their work.
> how is more work being scheduled?

@yuvalif `io_context.stop()` just causes `io_context.run()` to return early so the worker threads can join. it doesn't wait for the `Manager::process_queues()` coroutine to complete, it just stops running it

the notify manager's coroutines issue rados requests that complete on its `io_context` (to be specific, on a strand executor wrapping that `io_context` - that's why `strand_executor_service` shows up in the valgrind report). if there are outstanding rados requests when Manager shuts down, the rados client will still try to schedule their completions on that strand executor

**Comment by mattbenjamin** [comment]
jenkins test api

File: src/rgw/rgw_appmain.cc [hunk 13]
```diff
@@ -0,0 +1,608 @@
...
+    if (r < 0) {
+      derr << "ERROR: failed run" << dendl;
+      return -r;
+    }
+
+    fes.push_back(fe);
+  }
+
+  std::string daemon_type = (nfs) ? "rgw-nfs" : "nfs";
+  r = store->register_to_service_map(dpp, daemon_type, service_map_meta);
```
Comment (epuertat) [inline src/rgw/rgw_appmain.cc:481]: @mattbenjamin Could this be the issue? The Dashboard looks for `rgw` daemons in the service map.

Comment (mattbenjamin) [reply src/rgw/rgw_appmain.cc]: as agreed earlier, I assume so; thanks for spotting this! (this still doesn't belong in jenkins at all)

**Comment by mattbenjamin** [comment]
merging while the merging is good!

**Comment by ivancich** [comment]
@mattbenjamin What's the difference semantically between init_frontends1 and init_frontends2? I realize there's an ordering, but what determines in which a specific frontend should go?

**Comment by mattbenjamin** [comment]
> @mattbenjamin What's the difference semantically between init_frontends1 and init_frontends2? I realize there's an ordering, but what determines in which a specific frontend should go?

Hi Eric.  The main distinction is, we discover each configured frontend, and allocate a FrontendConfig object, in init_frontends1().  But we do not initialize the frontends until init_frontends2().  In between, various additional actions are taken, which I carried forward--some of them could almost certainly move around, but this order works, and is carried forward from the ordering of actions in rgw's main(...) function as it's existed for years...

**Pull request merged**